designs/comprehensions-design.md

```diff
@@ -0,0 +1,216 @@
+
+
+## Workflow Details:
+
+Since, we have not yet collaborated with the lists group, we are only printing the elements of the list (constructed by the comprehension expression). We plan to collaborate with the lists team in the upcoming week, and display the lists in the proper format. For example:
+
+```print([1 for a in range(1,5)])``` gives the following output instead of ```[1,1,1,1,1]```::
+```
+1
+1
+1
+1
+1
+```
+We have added two new fields to the ast structure for list-comp: iterable_cond? and body. iterable condition checks if range.hasNext() returns true or not. body stores the following statements:: 1. print the current element, 2. call the next method. In the lower.ts file, we have converted the ast structure into an array of basic blocks. Also, as of this week, we are explicitly adding a range class in our test cases, but we would remove this as part of our next milestones. 
+
+
+
+
+## Week 6:
+## Test cases
+
+```
+a : List = None
+a = [i for i in range(10)]
+print(a)
+```
+This should output ```[0,1,2,3,4,5,6,7,8,9]```
+
+```
+a : List = None
+j : int = 9
+a = [j for i in range(5)]
+print(a)
+```
+This should output ```[9,9,9,9,9]```
+
+```
+a : List = None
+a = [j for i in range(10)]
+print(a)
+```
+This should output ```Error: j is not defined```
+
+```
+a : List = None
+a = [i for i in range(5,10)]
+print(a)
+```
+This should output ```[5,6,7,8,9]```
+
+```
+a : List = None
+a = [i for i in range(10,5)]
+print(a)
+```
+This should output ```Error: list range is invalid```
+
+```
+a : List = None
+a = [i for i in range(10) if i < 5]
+print(a)
+```
+This should output ```[0,1,2,3,4]```
+
+```
+a : List = None
+a = [i for i in range(10) return]
+print(a)
+```
+This should output ```Error: only if condition allowed```
+
+```
+a : List = None
+b : List = None
+a = [1,2,3]
+b = [i*2 for i in a]
+print(b)
+```
+This should output ```[2,4,6]```
+
+```
+a : List = None
+a = [[j for j in range(3)] for i in range(3)]
+print(a)
+```
+This should output ```[[0, 1, 2], [0, 1, 2], [0, 1, 2]]```
+
+```
+a : List = None
+b : List = None
+a = [1,2,3]
+j : bool = True
+b = [j for i in a]
+print(b)
+```
+This should output ```[True,True,True]```
+
+```
+a : int = 5
+b : List = None
+b = [i for i in a]
+print(b)
+```
+This should output ```Error: variable a is not an iterable```
+
+## Code changes
+
+We plan to have an external class Iterable for now with corresponding ```next()``` and ```hasNext()``` methods. Every time a list comprehension will be done, a new instance of this class will be created and used to assign values to the new list variable. We want to create a seperate file to construct the range class under the directory models. We can talk about the range class in more details after we have a discussion with the lists group and once we start our implementation.
+
+In the ast.ts file,  we plan to add a new Type for lists:
+```
+{ tag: "list", listitems : Array<Type>}
```

Comment:
This really needs to be added as soon as possible.

Reply:
Regarding this, in week 1 we got a review which advised us to wait for the lists team to make this change. We plan on utilizing the change made by the lists team as part of our next milestone.